BBQJIT should support gc and funcref opcodes

kmiller68 opened the pull request:
#### 6ef70e7d7c5a168306c05061320d37496bf285e0
<pre>
BBQJIT should support gc and funcref opcodes
<a href="https://bugs.webkit.org/show_bug.cgi?id=256959">https://bugs.webkit.org/show_bug.cgi?id=256959</a>

Reviewed by Yusuke Suzuki.

This patch adds support for the various gc and funcref opcodes to the new BBQ JIT.
Most of the implementations are just translations of what the B3IRGenerator does.
The main difference is that for opcodes which need to make a C call, e.g. for allocation,
they do so by creating a `Value::fromTemp` that does not conflict with any parameter `Value`.
This is needed because otherwise the BBQJIT allocator gets confused between the existing parameters
that were not passed to the C call and the result of the C call. Also, since BBQJIT doesn&apos;t have a
good way to branch over a call both `ref.cast` and `ref.test` just call an operation.

Also, this patch fixes an issue where we weren&apos;t checking for the spec&apos;s limit on array.new_fixed
static argument count.

Lastly, there is a workaround for a clang bug where it crashed when compiling a unified source.
The workaround was to @no-unify one of the files in that bundle.

* JSTests/wasm/gc/array_new_fixed.js:
* JSTests/wasm/gc/i31.js:
(testI31Get):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::isValidValueTypeKind):
(JSC::Wasm::BBQJIT::pointerType):
(JSC::Wasm::BBQJIT::toValueKind):
(JSC::Wasm::BBQJIT::Value::fromRef):
(JSC::Wasm::BBQJIT::Value::fromTemp):
(JSC::Wasm::BBQJIT::Value::fromLocal):
(JSC::Wasm::BBQJIT::Value::pinned):
(JSC::Wasm::BBQJIT::Value::type const):
(JSC::Wasm::BBQJIT::BBQJIT):
(JSC::Wasm::BBQJIT::addTableGet):
(JSC::Wasm::BBQJIT::addTableSet):
(JSC::Wasm::BBQJIT::addTableInit):
(JSC::Wasm::BBQJIT::addTableSize):
(JSC::Wasm::BBQJIT::addTableGrow):
(JSC::Wasm::BBQJIT::addTableFill):
(JSC::Wasm::BBQJIT::addTableCopy):
(JSC::Wasm::BBQJIT::addGrowMemory):
(JSC::Wasm::BBQJIT::addMemoryFill):
(JSC::Wasm::BBQJIT::addMemoryCopy):
(JSC::Wasm::BBQJIT::addMemoryInit):
(JSC::Wasm::BBQJIT::atomicWait):
(JSC::Wasm::BBQJIT::atomicNotify):
(JSC::Wasm::BBQJIT::addI31New):
(JSC::Wasm::BBQJIT::addI31GetS):
(JSC::Wasm::BBQJIT::addI31GetU):
(JSC::Wasm::BBQJIT::getTypeDefinition):
(JSC::Wasm::BBQJIT::getArrayTypeDefinition):
(JSC::Wasm::BBQJIT::getArrayElementType):
(JSC::Wasm::BBQJIT::marshallToI64):
(JSC::Wasm::BBQJIT::addArrayNew):
(JSC::Wasm::BBQJIT::addArrayNewDefault):
(JSC::Wasm::BBQJIT::pushArrayNewFromSegment):
(JSC::Wasm::BBQJIT::addArrayNewData):
(JSC::Wasm::BBQJIT::addArrayNewElem):
(JSC::Wasm::BBQJIT::emitArraySetUnchecked):
(JSC::Wasm::BBQJIT::addArrayNewFixed):
(JSC::Wasm::BBQJIT::addArrayGet):
(JSC::Wasm::BBQJIT::addArraySet):
(JSC::Wasm::BBQJIT::addArrayLen):
(JSC::Wasm::BBQJIT::emitStructSet):
(JSC::Wasm::BBQJIT::addStructNewDefault):
(JSC::Wasm::BBQJIT::addStructNew):
(JSC::Wasm::BBQJIT::addStructGet):
(JSC::Wasm::BBQJIT::addStructSet):
(JSC::Wasm::BBQJIT::addRefTest):
(JSC::Wasm::BBQJIT::addRefCast):
(JSC::Wasm::BBQJIT::addExternInternalize):
(JSC::Wasm::BBQJIT::emitThrowOnNullReference):
(JSC::Wasm::BBQJIT::addI32Popcnt):
(JSC::Wasm::BBQJIT::addI64Popcnt):
(JSC::Wasm::BBQJIT::addRefFunc):
(JSC::Wasm::BBQJIT::toB3Type):
(JSC::Wasm::BBQJIT::emitCCall):
(JSC::Wasm::BBQJIT::addCallRef):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
* Source/JavaScriptCore/wasm/WasmLimits.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:

Canonical link: <a href="https://commits.webkit.org/264638@main">https://commits.webkit.org/264638@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/450b85be1377533841691977c54b846572249f1b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/8185 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/8476 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/8693 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/9848 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/8255 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/10464 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/8387 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/11127 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/8330 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/9395 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/7430 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/9970 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/6695 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/7489 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/15048 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/6979 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/7818 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/7619 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/10975 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/7751 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/8091 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/6576 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/8357 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/7383 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1922 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1977 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/11590 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/8581 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/7833 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2068 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -385,6 +403,39 @@ class BBQJIT {
+        case TypeKind::Eqref:
+        case TypeKind::Anyref:
+        case TypeKind::Nullref:
+            // FIXME: Should this just be I64?
```
I think they are I64 since it is EncodedJSValue (it can also be jsNull etc.).

kmiller68 replied:
Yeah, I can do that.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3585,27 +3643,561 @@ class BBQJIT {
+        return { };
+    }
+
+    void emitStructSet(Value structValue, const StructType& structType, uint32_t fieldIndex, Value value)
```
I wonder if we should pass `GPRReg structValueGPR` here since it is confusing that we consume value while we do not consume structValue.

kmiller68 replied:
Done.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3585,27 +3643,561 @@ class BBQJIT {
+        }
+
+        Location arrayLocation = loadIfNecessary(arrayref);
+        emitThrowOnNullReference(ExceptionType::NullArrayLen, arrayLocation);
```
Dont we need to consume arrayref?

kmiller68 replied:
Yes, fixed. I think it kinda works right now because result is the same temp as array and gets the same location. I was trying to think of a validation for this but I couldn't come up with anything that would obviously work.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3585,27 +3643,561 @@ class BBQJIT {
+            return { };
+        }
+
+        Location arrayLocation = loadIfNecessary(arrayref);
```
Don't we need to consume arrayref?

kmiller68 replied:
No, it get's consumed by the emitCCall in `emitArraySetUnchecked`.

ddegazio commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3585,27 +3642,563 @@ class BBQJIT {
+
+        LOG_INSTRUCTION("I31New", value, RESULT(result));
+
+        // Should I be able to load this from an Address? Doesn't matter on ARM tho.
```
On this, right now we tend to just materialize constants inline throughout BBQJIT. We don't have the luxury of lifting shared constants or knowing if a function will ever use i.e. `NumberTag` (hence we'd have to conservatively always reserve a slot for it). So this is consistent with how other constants are handled in BBQJIT, although we could definitely consider improving this in the future.

kmiller68 replied:
AH, yeah I should have deleted that comment. I wrote that when as part of the first op I implemented so I didn't know what I was doing.

ddegazio commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3585,27 +3642,563 @@ class BBQJIT {
+            Value::fromI32(typeIndex),
+            Value::fromI32(args.size()),
+        };
+        Value allocationResult = Value::fromTemp(TypeKind::Structref, currentControlData().enclosedHeight() + currentControlData().implicitSlots() + m_parser->expressionStack().size() + args.size());
```
Why is this `Structref` and not `Arrayref`? Or `I64` for that matter (since `operationWasmArrayNewEmpty` returns an encoded JSValue)?

kmiller68 replied:
Copy paste error, will fix.

ddegazio commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3585,27 +3642,563 @@ class BBQJIT {
+            emitArraySetUnchecked(typeIndex, pinnedResult, Value::fromI32(i), args[i]);
+        }
+
+        result = topValue(TypeKind::Structref);
```
Ditto.

kmiller68 replied:
Fixed.

ddegazio commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -385,6 +403,38 @@ class BBQJIT {
         return 0;
     }
 
+    static TypeKind toValueKind(TypeKind kind)
```
If we're going to start caring about distinguishing between BBQ value kind and WASM type, maybe this is the time to define a new enum `ValueKind` for the former instead of using `TypeKind` in the current ad-hoc way. Would make it a lot clearer what a given type is being used for at any given time.

kmiller68 replied:
Yeah, I think I'll do that in a different patch though.

ddegazio commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -385,6 +403,38 @@ class BBQJIT {
+        case TypeKind::Eqref:
+        case TypeKind::Anyref:
+        case TypeKind::Nullref:
+            return TypeKind::I64;
```
Maybe we can define an alias for `TypeKind::I64` called something like `TypeKind::JSValue` (or preferably, we put this in some new `ValueKind` enumeration). Functionally this wouldn't change anything (and I guess we don't go this route in B3), but looking at the instruction implementations below it could be helpful for following whether we're using an I64 as an arithmetic integer v.s. as a means for storing a JS value. Maybe also enables better assertions and logging.

kmiller68 replied:
Yeah, that'd be a good idea.

ddegazio commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -3585,27 +3642,563 @@ class BBQJIT {
+
+        if (isFloatingPointType(resultType.kind)) {
+            consume(getResult);
+            result = Value::fromTemp(resultType.kind, currentControlData().enclosedHeight() + currentControlData().implicitSlots() + m_parser->expressionStack().size());
```
I think this can just be `result = topValue(resultType.kind);`

kmiller68 replied:
Changed.

ddegazio commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -7171,10 +7774,10 @@ class BBQJIT {
     {
         // Currently, we assume the Wasm calling convention is the same as the C calling convention
-        Vector<Type, 1> resultTypes = { Type { returnType, 0u } };
+        Vector<Type, 1> resultTypes = { Type { result.type(), 0u } };
```
Let's add an `ASSERT(!result.isNone());` to make sure nobody accidentally passes in an empty value.

kmiller68 replied:
There's an ASSERT(result.isTemp()) below. I'll move that to the top of the function.

ddegazio commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5449,8 +6047,10 @@ class BBQJIT {
-                else
-                    emitCCall(&operationPopcount32, Vector<Value> { operand }, TypeKind::I32, result);
+                else {
+                    result = topValue(TypeKind::I32);
```
The `EMIT_UNARY(...)` macro will already assign `result` to the top value on the stack and give it a register, so it should be able to be passed in. However, this creates a somewhat nasty tacit dependency - `emitCCall` will bind `result` to the `returnValueGPR`, which will error if `result` is already bound to a different register (to avoid mucking with the register allocator state). It shouldn't currently error specifically because we only allocate caller-saved registers, which get flushed across the call regardless; however, if we add callee-saved register allocation to BBQJIT in the future, we could get very niche errors.

All this is to say, I think the most correct thing with the current abstractions would be to
```
consume(result);
emitCCall(&operationPopcount32, Vector<Value> { operand }, result);
```
so `result` is quietly unbound from its current register before we pass it into `emitCCall`. This only applies to `addI32Popcnt` and `addI64Popcnt` since they are currently the only functions that use `EMIT_UNARY`/`EMIT_BINARY` and also contain a call.

kmiller68 replied:
I'll make the change to consume. Unfortunately I think we won't have any bots that test this code path anyway. Since I don't think we have any bots that run on old enough CPUs that they'll fail the `supportsCountPopulation()` test now that arm64 always passes. (For reference the first Intel processor with POPCNT came out 14 years ago https://en.wikipedia.org/wiki/Nehalem_(microarchitecture) and I think earlier for AMD)

ddegazio commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5463,8 +6063,10 @@ class BBQJIT {
-                else
-                    emitCCall(&operationPopcount64, Vector<Value> { operand }, TypeKind::I32, result);
+                else {
+                    result = topValue(TypeKind::I32);
```
This isn't strictly related to your patch but I just realized this is slightly wrong...`operationPopcount64` returns a 64-bit result, not a 32-bit one, so `I32` is the wrong type. If we reuse the `result` from `EMIT_UNARY` like I mentioned in my previous comment, it should have the correct type already; otherwise, this should probably use `TypeKind::I64`.

kmiller68 replied:
Fixed via switching it to `consume`.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5449,8 +6046,10 @@ class BBQJIT {
-                else
-                    emitCCall(&operationPopcount32, Vector<Value> { operand }, TypeKind::I32, result);
+                else {
+                    consume(result);
```
Can we have explicit detailed comment about this hack here?

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -5463,8 +6062,10 @@ class BBQJIT {
-                else
-                    emitCCall(&operationPopcount64, Vector<Value> { operand }, TypeKind::I32, result);
+                else {
+                    consume(result);
```
Ditto.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -7394,7 +7996,43 @@ class BBQJIT {
+        GPRReg calleeInstance, calleeCode, jsCalleeAnchor;
+
+        {
+            ScratchScope<3, 0> scratches(*this);
```
Please update this code to the latest convention. `calleeCode` always needs to be in GPRInfo::nonPreservedNonArgumentGPR1 to make shuffling work correctly. See updated `addCallIndirect` code changed in https://github.com/WebKit/WebKit/commit/76f47aa9042386307ab0e61744870833fd981a76

Constellation reviewed (approved):
r=me with comments and update of `addCallRef`.

Constellation commented on Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
```diff
@@ -7409,7 +8027,47 @@ class BBQJIT {
+        GPRReg calleeInstance;
+        GPRReg calleeCode;
+        GPRReg jsCalleeAnchor;
+        {
```
You need `clobber(GPRInfo::nonPreservedNonArgumentGPR1)`

kmiller68 committed:
BBQJIT should support gc and funcref opcodes
https://bugs.webkit.org/show_bug.cgi?id=256959

Reviewed by Yusuke Suzuki.

This patch adds support for the various gc and funcref opcodes to the new BBQ JIT.
Most of the implementations are just translations of what the B3IRGenerator does.
The main difference is that for opcodes which need to make a C call, e.g. for allocation,
they do so by creating a `Value::fromTemp` that does not conflict with any parameter `Value`.
This is needed because otherwise the BBQJIT allocator gets confused between the existing parameters
that were not passed to the C call and the result of the C call. Also, since BBQJIT doesn't have a
good way to branch over a call both `ref.cast` and `ref.test` just call an operation.

Also, this patch fixes an issue where we weren't checking for the spec's limit on array.new_fixed
static argument count.

Lastly, there is a workaround for a clang bug where it crashed when compiling a unified source.
The workaround was to @no-unify one of the files in that bundle.

* JSTests/wasm/gc/array_new_fixed.js:
* JSTests/wasm/gc/i31.js:
(testI31Get):
* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/Options.cpp:
(JSC::Options::notifyOptionsChanged):
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJIT::isValidValueTypeKind):
(JSC::Wasm::BBQJIT::pointerType):
(JSC::Wasm::BBQJIT::toValueKind):
(JSC::Wasm::BBQJIT::Value::fromRef):
(JSC::Wasm::BBQJIT::Value::fromTemp):
(JSC::Wasm::BBQJIT::Value::fromLocal):
(JSC::Wasm::BBQJIT::Value::pinned):
(JSC::Wasm::BBQJIT::Value::type const):
(JSC::Wasm::BBQJIT::BBQJIT):
(JSC::Wasm::BBQJIT::addTableGet):
(JSC::Wasm::BBQJIT::addTableSet):
(JSC::Wasm::BBQJIT::addTableInit):
(JSC::Wasm::BBQJIT::addTableSize):
(JSC::Wasm::BBQJIT::addTableGrow):
(JSC::Wasm::BBQJIT::addTableFill):
(JSC::Wasm::BBQJIT::addTableCopy):
(JSC::Wasm::BBQJIT::addGrowMemory):
(JSC::Wasm::BBQJIT::addMemoryFill):
(JSC::Wasm::BBQJIT::addMemoryCopy):
(JSC::Wasm::BBQJIT::addMemoryInit):
(JSC::Wasm::BBQJIT::atomicWait):
(JSC::Wasm::BBQJIT::atomicNotify):
(JSC::Wasm::BBQJIT::addI31New):
(JSC::Wasm::BBQJIT::addI31GetS):
(JSC::Wasm::BBQJIT::addI31GetU):
(JSC::Wasm::BBQJIT::getTypeDefinition):
(JSC::Wasm::BBQJIT::getArrayTypeDefinition):
(JSC::Wasm::BBQJIT::getArrayElementType):
(JSC::Wasm::BBQJIT::marshallToI64):
(JSC::Wasm::BBQJIT::addArrayNew):
(JSC::Wasm::BBQJIT::addArrayNewDefault):
(JSC::Wasm::BBQJIT::pushArrayNewFromSegment):
(JSC::Wasm::BBQJIT::addArrayNewData):
(JSC::Wasm::BBQJIT::addArrayNewElem):
(JSC::Wasm::BBQJIT::emitArraySetUnchecked):
(JSC::Wasm::BBQJIT::addArrayNewFixed):
(JSC::Wasm::BBQJIT::addArrayGet):
(JSC::Wasm::BBQJIT::addArraySet):
(JSC::Wasm::BBQJIT::addArrayLen):
(JSC::Wasm::BBQJIT::emitStructSet):
(JSC::Wasm::BBQJIT::addStructNewDefault):
(JSC::Wasm::BBQJIT::addStructNew):
(JSC::Wasm::BBQJIT::addStructGet):
(JSC::Wasm::BBQJIT::addStructSet):
(JSC::Wasm::BBQJIT::addRefTest):
(JSC::Wasm::BBQJIT::addRefCast):
(JSC::Wasm::BBQJIT::addExternInternalize):
(JSC::Wasm::BBQJIT::emitThrowOnNullReference):
(JSC::Wasm::BBQJIT::addI32Popcnt):
(JSC::Wasm::BBQJIT::addI64Popcnt):
(JSC::Wasm::BBQJIT::addRefFunc):
(JSC::Wasm::BBQJIT::toB3Type):
(JSC::Wasm::BBQJIT::emitCCall):
(JSC::Wasm::BBQJIT::addCallRef):
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::FunctionParser<Context>::parseExpression):
* Source/JavaScriptCore/wasm/WasmLimits.h:
* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperations.h:

Canonical link: https://commits.webkit.org/264638@main
